optimize_workspace.sh

```diff
@@ -18,29 +18,61 @@ cargo --version
 rm -f target/wasm32-unknown-unknown/release/*.wasm
 
 # Build artifacts
-echo "Building artifacts in workspace ..."
+echo "Building artifacts in workspace..."
 /usr/local/bin/build_workspace
 
 mkdir -p artifacts
-echo "Creating intermediate hashes ..."
-sha256sum -- target/wasm32-unknown-unknown/release/*.wasm | tee artifacts/checksums_intermediate.txt
-
-echo "Optimizing artifacts in workspace ..."
+echo "Optimizing artifacts in workspace..."
 TMPARTIFACTS=$(mktemp -p "$(pwd)" -d artifacts.XXXXXX)
 # Optimize artifacts
 (
   cd "$TMPARTIFACTS"
 
   for WASM in ../target/wasm32-unknown-unknown/release/*.wasm; do
-    NAME=$(basename "$WASM" .wasm)${SUFFIX}.wasm
-    echo "Optimizing $NAME ..."
-    wasm-opt -Os "$WASM" -o "$NAME"
+    BASENAME=$(basename "$WASM" .wasm)
+    NAME=${BASENAME}${SUFFIX}
+    OPTIMIZED_WASM=${NAME}.wasm
+
+    INTERMEDIATE_SHA=$(sha256sum -- "$WASM" | sed 's,../target,target,g')
+    INTERMEDIATE_SHAS="../artifacts/checksums_intermediate.txt"
+
+    SKIP_OPTIMIZATION=false
+    if test -f "../artifacts/${OPTIMIZED_WASM}"; then
+      INTERMEDIATE_CACHE_HIT=$(
+        grep -Fxsq "$INTERMEDIATE_SHA" "$INTERMEDIATE_SHAS"
+        echo $?
+      )
+      OPTIMIZED_SHA=$(sha256sum -- "../artifacts/$OPTIMIZED_WASM" | sed 's,../artifacts/,,g')
+      OPTIMIZED_SHAS="../artifacts/checksums.txt"
```

Comment:
Idem.

Reply:
...and this one has a dependency on `$OPTIMIZED_WASM` which has a dependency on `$WASM`

Reply:
Same as above. The constants can be moved out of the for loop.

Reply:
Fixed!